Post gui challenges building settings (#4855)
GoogleChrome/web.dev: src/site/content/en/blog/building-a-settings-component/index.md

authors:
  - adamargyle
description: A foundational overview of how to build a settings component of sliders and checkboxes.
date: 2021-03-17
hero: image/vS06HQ1YTsbMKSFTIPl2iogUQP73/SUaxDTgOYvv2JXxaErBP.png
tags:
  - blog
  - css
  - dom
  - javascript
  - layout
  - mobile
  - ux
---

In this post I want to share thinking on building a Settings component for the
web that is responsive, supports multiple device inputs, and works across
browsers. Try the [demo](https://gui-challenges.web.app/settings/dist/).

<figure class="w-figure w-figure--fullbleed">
  {% Video 
    src="video/vS06HQ1YTsbMKSFTIPl2iogUQP73/WuIwd9jPb30KmmnjJn75.mp4",
    className="w-screenshot", 
    autoplay="true", 
    loop="true", 
    muted="true" 
  %}
  <figcaption class="w-figure">
    <a href="https://gui-challenges.web.app/settings/dist/">Demo</a>
  </figcaption>
</figure>

If you prefer video, or want a UI/UX preview of what we're building, here's a
shorter walkthrough on YouTube:

{% YouTube 'mMBcHcvxuuA' %}

## Overview

I've broken out the aspects of this component into the following sections:

1. [Layouts](#layouts)
1. [Color](#color)

> 1. [Custom Range Input](#custom-range)
> 1. [Custom Checkbox Input](#custom-checkbox)
> 1. [Accessibility Considerations](#accessibility)

Reviewer: ```suggestion
1. [Custom range input](#custom-range)
1. [Custom checkbox input](#custom-checkbox)
1. [Accessibility considerations](#accessibility)
```

1. We passed an extra `min()` function
1. We specify `align-items: flex-start`
1. There's a `max-width: 89vw` style?

The extra `min()` function is well described by Evan Minto on their blog in the
post [Intrinsically Responsive CSS Grid with minmax() and
min()](https://evanminto.com/blog/intrinsically-responsive-css-grid-minmax-min/).
I recommend giving that a read. The `flex-start` alignment correction is to
remove the default stretching effect, so that the children of this layout don't
need to have equal heights, they can have natural, intrinsic heights. The
YouTube video has a quick breakdown of this alignment addition.

`max-width: 89vw` is worth a small breakdown in this post. It's partly a
bug, and partly a learning moment. Let me show you the layout with and without
the style applied:

{% Video 
  src="video/vS06HQ1YTsbMKSFTIPl2iogUQP73/gdldf7hyaBrHWwxQbSaT.mp4",
  className="w-screenshot", 
  autoplay="true", 
  loop="true", 
  muted="true" 
%}

What's happening? When `max-width` is specified, it's providing context,
explicit sizing or [definite
sizing](https://drafts.csswg.org/css-sizing-3/#definite) for the [`auto-fit`

> layout algorythym](https://drafts.csswg.org/css-grid/#auto-repeat) to know how

Reviewer: ```suggestion
layout algorithm](https://drafts.csswg.org/css-grid/#auto-repeat) to know how
```

The trick to this is "revealing" the vibrant fill color. This is done with the
hard stop gradient on top, it's transparent up to the fill percentage, and after
that it's the unfilled track surface color. Behind that unfilled surface, is a
full width color, waiting for transparency to reveal it.

{% Video 
  src="video/vS06HQ1YTsbMKSFTIPl2iogUQP73/aiAL28AkDRZvaAZNEbW8.mp4",
  className="w-screenshot", 
  autoplay="true", 
  loop="true", 
  muted="true" 
%}

My solution did require Javascript though, in order to maintain the fill style:

```js
/* grab sliders on page */
const sliders = document.querySelectorAll('input[type="range"]')

/* take a slider element, return a percentage string for use in CSS */
const rangeToPercent = slider => {
  const max = slider.getAttribute('max') || 10;
  const percent = slider.value / max * 100;

  return `${parseInt(percent)}%`;
};

/* on page load, set the fill amount */
sliders.forEach(slider => {
  slider.style.setProperty('--track-fill', rangeToPercent(slider));

  /* when a slider changes, update the fill prop */
  slider.addEventListener('input', e => {
    e.target.style.setProperty('--track-fill', rangeToPercent(e.target));
  })
})
```

> I think this makes for a nice visual upgrade. The slider works great without
> Javascript, the `--track-fill` prop is not required. If it's important to the
> design though, server side components could set the property on the html,
> allowing Javascript to hydrate the value while also observing any user changes. 

Reviewer: ```suggestion
I think this makes for a nice visual upgrade. The slider works great without
Javascript, the `--track-fill` prop is not required. 
```

I think this needs an explanation of what the JS is doing, why you needed it.

There are easier to manage solutions for custom track fill colors, but the style
I wanted necessitated some Javascript. [Here's a great
post](https://css-tricks.com/sliding-nightmare-understanding-range-input/) on
[CSS-Tricks](https://css-tricks.com/) by [Ana
Tudor](https://twitter.com/anatudor), that demonstrates a CSS only solution for
track fill.

### Thumb styles

```css
input[type="range"]::-webkit-slider-thumb {
  appearance: none; /* clear styles, make way for mine */
  cursor: ew-resize; /* cursor style to support drag direction */
  border: 3px solid var(--surface3);
  block-size: var(--thumb-size);
  inline-size: var(--thumb-size);
  margin-top: var(--thumb-offset);
  border-radius: 50%;
  background: var(--brand-bg-gradient) fixed;
}
```

> The majority of these styles are around making a nice tangible looking circle
> element. Again you see the fixed background gradient there that unifies the
> dynamic colors of the thumbs, tracks and associated SVG elements.
> 

Reviewer: ```suggestion
The majority of these styles are to make a nice circle. 
Again you see the fixed background gradient there that unifies the
dynamic colors of the thumbs, tracks and associated SVG elements.
```

I separated the styles for the interaction to help isolate the `box-shadow`

> technique being user for the hover highlight:

Reviewer: ```suggestion
technique being used for the hover highlight:
```

```css
@custom-media --motionOK (prefers-reduced-motion: no-preference);

::-webkit-slider-thumb {
  …

  /* shadow spread is initally 0 */
  box-shadow: 0 0 0 var(--thumb-highlight-size) var(--thumb-highlight-color);
  
  /* if motion is OK, transition the box-shadow change */
  @media (--motionOK) {
    & {
      transition: box-shadow .1s ease;
    }
  }

  /* on hover/active state of parent, increase size prop */
  @nest input[type="range"]:is(:hover,:active) & {
    --thumb-highlight-size: 10px;
  }
}
```

> The goal was a nice painted highlight on the user interactive element. By using
> a box shadow that's not blurred and matches the circular shape of the thumb
> element, we can change and transition it's spread size, resulting in an easy to
> manage visual highlight effect. 

Reviewer: ```suggestion
The goal was an east to manage painted highlight on the user interactive element. By using
a box shadow that's not blurred and matches the circular shape of the thumb
element, we can change and transition it's spread size. 
```

{% Video 
  src="video/vS06HQ1YTsbMKSFTIPl2iogUQP73/s835RbH88L5bxjl5bMFl.mp4",
  className="w-screenshot", 
  autoplay="true", 
  loop="true", 
  muted="true" 
%}

If only the highlight effect was so easy on checkboxes…

### Cross browser selectors

I found I needed these `-webkit-` and `-moz-` selectors to achieve cross browser
consistency:

```css
input[type="range"] {
  &::-webkit-slider-runnable-track {}
  &::-moz-range-track {}
  &::-webkit-slider-thumb {}
  &::-moz-range-thumb {}
}
```

{% Aside 'gotchas' %} 
[Josh Comeau](https://twitter.com/JoshWComeau) outlines
why the above examples don't simply use a comma between selectors for cross
browser styling, see the [Twitter
thread](https://twitter.com/JoshWComeau/status/1359213591602335752?s=20) for
more information. 
{% endAside %}

## Custom Checkbox

> Given the following HTML input element, I'll show you how I customized it's

Reviewer: ```suggestion
Given the following HTML input element, I'll show you how I customized its
```

```

There's 3 parts to this element we need to customize:
1. [Checkbox element](#checkbox-element)
1. [Associated labels](#checkbox-labels)
1. [Highlight effect](#checkbox-highlight)

### Checkbox element

```css
input[type="checkbox"] {
  inline-size: var(--space-sm);   /* increase width */
  block-size: var(--space-sm);    /* increase height */
  outline-offset: 5px;            /* focus style enhancement */
  accent-color: var(--brand);     /* tint the input */
  position: relative;             /* prepare for an absolute pseudo element */
  transform-style: preserve-3d;   /* create a 3d z-space stacking context */
  margin: 0;
  cursor: pointer;
}
```

Besides the pseudo element `transform-style` and `position` styles, it's mostly
minor opinionated style stuff from me. I like the cursor to be pointer, I like
outline offsets, default checkboxes are too tiny, and if `accent-color` is
[supported](https://drafts.csswg.org/css-ui-4/#widget-accent), bring these
checkboxes into the brand color scheme.

### Checkbox labels

It's important to provide labels for checkboxes for 2 reasons. The first is to
represent what the checkbox value is used for, to answer "on or off for what?".

> Second is for UX, web user's have become accustomed to interacting with

Reviewer: ```suggestion
Second is for UX, web users have become accustomed to interacting with
```

after I fixed it:

{% Video 
  src="video/vS06HQ1YTsbMKSFTIPl2iogUQP73/Spdpw5P1MD8ceazneRXo.mp4",
  className="w-screenshot", 
  autoplay="true", 
  loop="true", 
  muted="true" 
%}

It's definitely a microinteraction, but important to me to keep the visual
consistency. The animation scaling technique is the same as we've been using in
other places, we set a custom property to a new value and let CSS transition it
based on motion preferences, but it's special feature is `translateZ(-1px)`. The
parent created a 3D space and this speudo element child tapped into it by
placing itself slightly back in z-space.

## Accessibility

The YouTube video does a great demonstration of the mouse, keyboard and
screenreader interactions for this settings component. I'll call out some of the
details here.

### HTML Element Choices

```html
<form>
<header>
<fieldset>
<picture>
<label>
<input>
```

Each of these holds hints and tips to the user's browsing tool. Some elements
provide interaction hints, some connect interactivity, and some help shape the
acccessibility tree that a screenreader navigates. 

### HTML Attributes

> The icon next to the slider in my form, that's not useful to a screenreader,
> let's help it hide that from the flow. The label is also not needed for
> screenreaders, let's remove it from that flow as well:

Reviewer: ```suggestion
We can hide elements that are not needed by screenreaders, 
in this case the icon next to the slider and the label:
```

feels like maybe this needs a reference as to why they are not needed.

}
```

I call this layout "just for gap" because it only uses grid to add gaps between blocks. 

`5` layouts use this strategy, here's all of them displayed:

{% Img 
  src="image/vS06HQ1YTsbMKSFTIPl2iogUQP73/zYWSVLzdtrh1K8p8yUuA.png", 
  alt="Vertical grid layouts highlighted with outlines and filled in gaps", 
  width="800", 
  height="568" 
%}

The `fieldset` element, which contains each input group (`.fieldset-item`), is using `gap: 1px` to
create the hairline borders between elements. No tricky border solution!

<div class="w-columns">
{% Compare 'better', 'Filled gap' %}

```css
.grid {
  display: grid;
  gap: 1px;
  background: var(--bg-surface-1);

  & > .fieldset-item {
    background: var(--bg-surface-2);
  }
}
```

{% endCompare %}

{% Compare 'worse', 'Border trick' %}

```css
.grid {
  display: grid;

  & > .fieldset-item {
    background: var(--bg-surface-2);
    
    &:not(:last-child) {
      border-bottom: 1px solid var(--bg-surface-1);
    }
  }
}
```
{% endCompare %}
</div>

### Natural grid wrapping

The most complex layout ended up being the macro layout, the logical layout
system between `<main>` and `<form>`. 

#### Centering wrapping content

> A common learning moment with CSS layout is distinguishing centering of `items`
> vs `content`. Flexbox and grid both provide abilities to `align-items` or
> `align-content`, and when dealing with wrapping elements, `content` layout
> alignments will distribute space amongst the children as a group. 

Reviewer: ```suggestion
Flexbox and grid both provide abilities to `align-items` or
`align-content`, and when dealing with wrapping elements, `content` layout
alignments will distribute space amongst the children as a group. 
```

```css
main {
  display: grid;
  gap: var(--space-xl);
  place-content: center;
}
```

> The main element is using `place-content: center` [alignment
> shorthand](https://developer.mozilla.org/en-US/docs/Web/CSS/place-content) so
> that regardless of a 2 or 1 column layout, the children should be centered
> vertically and horizontally, together as a group.

Reviewer: ```suggestion
The main element is using `place-content: center` [alignment
shorthand](https://developer.mozilla.org/en-US/docs/Web/CSS/place-content) so
that the children are centered vertically and horizontally in both one and two column layouts.
```

Steiner](https://web.dev/authors/thomassteiner/). There's a lot more to gain
than dark checkbox inputs!

### CSS `accent-color`

There's been [recent
activity](https://twitter.com/argyleink/status/1360022120810483715?s=20) around
`accent-color` on form elements, being a single CSS style that can change the
tint color used in the browsers input element. Read more about it [here on
GitHub](https://github.com/w3c/csswg-drafts/issues/5187). I've included it in my
styles for this component. As browsers support it, my checkboxes will be
more on theme with the pink and purple color pops.

```css
input[type="checkbox"] {
  accent-color: var(--brand);
}
```

{% Img 
  src="image/vS06HQ1YTsbMKSFTIPl2iogUQP73/J9pbhB0ImoDzbsXkBGtG.png", 
  alt="A screenshot from Chromium on Linux of pink checkboxes", 
  width="800", 
  height="406" 
%}

### Color pops with fixed gradients and focus-within

Color pops most when it's used sparingly, and one of the ways I like to achieve
that is through colorful UI interactions. 

{% Video 
  src="video/vS06HQ1YTsbMKSFTIPl2iogUQP73/Pm75QwVToKkiqedqPtmm.mp4",
  className="w-screenshot", 
  autoplay="true", 
  loop="true", 
  muted="true",
  width="480px"
%}

> 1. Highlight context
> 1. UI feedback of "how full" the value is in the range
> 1. UI feedback that it's accepting input

Reviewer: ```suggestion
There are many layers of UI feedback and interaction in the above video, which
help give personality to the interaction by:
- Highlighting context.
- Providing UI feedback of "how full" the value is in the range.
- Providing UI feedback that a field is accepting input.
```

> There are many layers of UI feedback and interaction in the above video, which
> help give personality to the interaction. Here, CSS is using the
> [`:focus-within`](https://developer.mozilla.org/en-US/docs/Web/CSS/:focus-within)
> pseudo class to change the appearance of various elements, let's break down the
> `.fieldset-item`, it's super interesting:

Reviewer: ```suggestion
To provide feedback when an element is being interacted with, CSS is using the
[`:focus-within`](https://developer.mozilla.org/en-US/docs/Web/CSS/:focus-within)
pseudo class to change the appearance of various elements, let's break down the
`.fieldset-item`, it's super interesting:
```

```css
.fieldset-item {
  ...
  
  &:focus-within {
    background: var(--surface2);

    & svg {
      fill: white;
    }

    & picture {
      clip-path: circle(50%);
      background: var(--brand-bg-gradient) fixed;
    }
  }
}
```

> When this element has focus is within of one of it's children:
> 1. `.fieldset-item` background is assigned a higher contrast surface color
> 1. The nested `svg` is filled white for higher contrast
> 1. The nested `<picture>` `clip-path` expands to a full circle and the
>    background is filled with the bright fixed gradient

Reviewer: ```suggestion
When one of the children of this element has focus-within:
1. The `.fieldset-item` background is assigned a higher contrast surface color.
1. The nested `svg` is filled white for higher contrast.
1. The nested `<picture>` `clip-path` expands to a full circle and the
   background is filled with the bright fixed gradient.
```

  border-radius: 50%;
  background: var(--brand-bg-gradient) fixed;
}
```

The majority of these styles are to make a nice circle. 
Again you see the fixed background gradient there that unifies the
dynamic colors of the thumbs, tracks and associated SVG elements.
I separated the styles for the interaction to help isolate the `box-shadow`
technique being used for the hover highlight:

```css
@custom-media --motionOK (prefers-reduced-motion: no-preference);

::-webkit-slider-thumb {
  …

  /* shadow spread is initally 0 */
  box-shadow: 0 0 0 var(--thumb-highlight-size) var(--thumb-highlight-color);
  
  /* if motion is OK, transition the box-shadow change */
  @media (--motionOK) {
    & {
      transition: box-shadow .1s ease;
    }
  }

  /* on hover/active state of parent, increase size prop */
  @nest input[type="range"]:is(:hover,:active) & {
    --thumb-highlight-size: 10px;
  }
}
```

{% Aside 'key-term' %} 
[@custom-media](https://drafts.csswg.org/mediaqueries-5/#custom-mq) 
is a Level 5 spec addition that
[PostCSS Custom Media](https://github.com/postcss/postcss-custom-media), part of
[PostCSS Preset Env](https://preset-env.cssdb.org/features). 
{% endAside %}

> The goal was an easy to manage and animated visual highlight for user feedback.
> By using a box shadow I can prevent effecting layout with the effect. I do this

Reviewer: ```suggestion
The goal was an easy to manage and animated visual highlight for user feedback.
By using a box shadow I can avoid triggering layout with the effect. I do this
```

Reviewer: might be good to add a link about what this means, maybe https://web.dev/animations-guide/ or there might be something you know of that is more focussed.

  &::-moz-range-track {}
  &::-webkit-slider-thumb {}
  &::-moz-range-thumb {}
}
```

{% Aside 'gotchas' %} 
[Josh Comeau](https://twitter.com/JoshWComeau) outlines
why the above examples don't simply use a comma between selectors for cross
browser styling, see the [Twitter
thread](https://twitter.com/JoshWComeau/status/1359213591602335752?s=20) for
more information. 
{% endAside %}

## Custom Checkbox

Given the following HTML input element, I'll show you how I customized its
appearance:

```html
<input type="checkbox">
```

There's 3 parts to this element we need to customize:
1. [Checkbox element](#checkbox-element)
1. [Associated labels](#checkbox-labels)
1. [Highlight effect](#checkbox-highlight)

### Checkbox element

```css
input[type="checkbox"] {
  inline-size: var(--space-sm);   /* increase width */
  block-size: var(--space-sm);    /* increase height */
  outline-offset: 5px;            /* focus style enhancement */
  accent-color: var(--brand);     /* tint the input */
  position: relative;             /* prepare for an absolute pseudo element */
  transform-style: preserve-3d;   /* create a 3d z-space stacking context */
  margin: 0;
  cursor: pointer;
}
```

> Besides the pseudo element `transform-style` and `position` styles, it's mostly
> minor opinionated style stuff from me. I like the cursor to be pointer, I like
> outline offsets, default checkboxes are too tiny, and if `accent-color` is
> [supported](https://drafts.csswg.org/css-ui-4/#widget-accent), bring these
> checkboxes into the brand color scheme.

Reviewer: ```suggestion
The `transform-style` and `position` styles prepare for the pseudo-element we will introduce later 
to style the highlight. Otherwise, it's mostly
minor opinionated style stuff from me. I like the cursor to be pointer, I like
outline offsets, default checkboxes are too tiny, and if `accent-color` is
[supported](https://drafts.csswg.org/css-ui-4/#widget-accent), bring these
checkboxes into the brand color scheme.
```

Is this edit correct? I couldn't figure out what you meant about the pseudo-element, but I think this is the ::before later on.

### Checkbox labels

It's important to provide labels for checkboxes for 2 reasons. The first is to

> represent what the checkbox value is used for, to answer "on or off for what?".

Reviewer: ```suggestion
represent what the checkbox value is used for, to answer "on or off for what?"
```

Second is for UX, web users have become accustomed to interacting with

> checkbox's via their associated labels.

Reviewer: ```suggestion
checkboxes via their associated labels.
```

{% Video 
  src="video/vS06HQ1YTsbMKSFTIPl2iogUQP73/7GYIFNjNCBdj13juFO7S.mp4",
  className="w-screenshot", 
  autoplay="true", 
  loop="true", 
  muted="true" 
%}

<div class="w-columns">
{% Compare 'better', 'input' %}

```html
<input 
  type="checkbox" 
  id="text-notifications" 
  name="text-notifications"
>
```

{% endCompare %}

{% Compare 'better', 'label' %}

```html
<label for="text-notifications">
  <h3>Text Messages</h3>
  <small>Get notified about all text messages sent to your device</small>
</label>
```
{% endCompare %}
</div>

On your label, put a `for` attribute that points to a checkbox by ID: `<label
for="text-notifications">`. On your checkbox, double up both the name and id to
ensure it's found with varying tools and tech, like a mouse or screenreader:
`<input type="checkbox" id="text-notifications" name="text-notifications">`.
`:hover`, `:active` and more come for free with the connection, increasing the
ways your form can be interacted with.

### Checkbox highlight

> I want to keep my interfaces consistent, and the slider element has a nice
> thumbnail highlight that I'd like to use with the checkbox. The thumbnail was
> able to use `box-shadow` and it's `spread` property to scale a shadow up and
> down, but that effect doesn't work here because our checkboxes are, [and should
> be!](https://twitter.com/argyleink/status/1329230409784291328?s=20), square.

Reviewer: ```suggestion
I want to keep my interfaces consistent, and the slider element has a nice
thumbnail highlight that I'd like to use with the checkbox. The thumbnail was
able to use `box-shadow` and it's `spread` property to scale a shadow up and
down. However, that effect doesn't work here because our checkboxes are, [and should
be](https://twitter.com/argyleink/status/1329230409784291328?s=20), square.
```

  block-size: var(--thumb-highlight-size);
  clip-path: circle(50%);                     /* circle shape */
  position: absolute;                         /* this is why position relative on parent */
  top: 50%;                                   /* pop and plop technique (https://web.dev/centering-in-css/#5.-pop-and-plop) */
  left: 50%;
  background: var(--thumb-highlight-color);
  transform-origin: center center;            /* goal is a centered scaling circle */
  transform:                                  /* order here matters!! */
    translateX(-50%)                          /* counter balances left: 50% */
    translateY(-50%)                          /* counter balances top: 50% */
    translateZ(-1px)                          /* PUTS IT BEHIND THE CHECKBOX */
    scale(var(--thumb-scale))                 /* value we toggle for animation */
  ;
  will-change: transform;

  @media (--motionOK) {                       /* transition only if motion is OK */
    & {
      transition: transform .2s ease;
    }
  }
}

/* on hover, set scale custom property to "in" state */
input[type="checkbox"]:hover::before {
  --thumb-scale: 1;
}
```

> Creating a circle psuedo element is straightforward work, but **placing it
> behind the element it's attached to**, that's what took work. Here's before and
> after I fixed it:

Reviewer: ```suggestion
Creating a circle psuedo-element is straightforward work, but **placing it
behind the element it's attached to** was harder. Here's before and
after I fixed it:
```

{% Video 
  src="video/vS06HQ1YTsbMKSFTIPl2iogUQP73/Spdpw5P1MD8ceazneRXo.mp4",
  className="w-screenshot", 
  autoplay="true", 
  loop="true", 
  muted="true" 
%}

> It's definitely a microinteraction, but important to me to keep the visual
> consistency. The animation scaling technique is the same as we've been using in
> other places, we set a custom property to a new value and let CSS transition it
> based on motion preferences, but it's special feature is `translateZ(-1px)`. The
> parent created a 3D space and this speudo element child tapped into it by
> placing itself slightly back in z-space.

Reviewer: ```suggestion
It's definitely a micro interaction, but important to me to keep the visual
consistency. The animation scaling technique is the same as we've been using in
other places. We set a custom property to a new value and let CSS transition it
based on motion preferences. The key feature here is `translateZ(-1px)`. The
parent created a 3D space and this pseudo-element child tapped into it by
placing itself slightly back in z-space.
```

## Accessibility

The YouTube video does a great demonstration of the mouse, keyboard and
screenreader interactions for this settings component. I'll call out some of the
details here.

### HTML Element Choices

```html
<form>
<header>
<fieldset>
<picture>
<label>
<input>
```

Each of these holds hints and tips to the user's browsing tool. Some elements
provide interaction hints, some connect interactivity, and some help shape the
acccessibility tree that a screenreader navigates. 

### HTML Attributes

We can hide elements that are not needed by screenreaders, 
in this case the icon next to the slider and the label:

```html
<picture aria-hidden="true">
<label aria-hidden="true">
```

{% Video 
  src="video/vS06HQ1YTsbMKSFTIPl2iogUQP73/fVjqHRZHQixAaxjeAvDP.mp4",
  className="w-screenshot", 
  autoplay="true", 
  loop="true", 
  muted="true",
  width="480px" 
%}

> The above video demonstrates the screenreader flow on Mac OS. Notice how input
> focus is straight from one slider to the next, this is because we've hidden the
> elements that may have otherwise been stops along the way to the next element.
> Without these attributes a user would need to stop, listen and move past the
> picture and duplicative label element.

Reviewer: ```suggestion
The above video demonstrates the screenreader flow on Mac OS. Notice how input
focus moves straight from one slider to the next. This is because we've hidden the
elements that may have become stops on the way to the next element.
Without these attributes a user would need to stop, listen and move past the
picture and duplicate label element.
```
